src/main/kotlin/com/getkeepsafe/dexcount/DexFile.kt

```diff
@@ -0,0 +1,292 @@
+import java.io.IOException
+import java.io.RandomAccessFile
+
+import java.util.concurrent.TimeUnit
+import java.util.zip.ZipEntry
+import java.util.zip.ZipException
+import java.util.zip.ZipFile
+
+
+/**
+ * A physical file and the {@link DexData} contained therein.
+ *
+ * A DexFile contains an open file, possibly a temp file.  When consumers are
+ * finished with the DexFile, it should be cleaned up with
+ * {@link DexFile#dispose()}.
+ */
+@Suppress("ConvertSecondaryConstructorToPrimary")
+internal class DexFile: Closeable {
+    private val file: File
+    private val isTemp: Boolean
+    private val raf: RandomAccessFile
+    private val data: DexData
+
+    val isInstantRun: Boolean
+
+    val methodRefs: List<MethodRef>
+        get() = data.methodRefs.toList()
+
+    val fieldRefs: List<FieldRef>
+        get() = data.fieldRefs.toList()
+
+    private constructor(file: File, isTemp: Boolean, isInstantRun: Boolean = false) {
+        this.file = file
+        this.isTemp = isTemp
+        this.isInstantRun = isInstantRun
+
+        this.raf = RandomAccessFile(file, "r")
+        this.data = DexData(raf)
+
+        try {
+            data.load()
+        } catch (e: IOException) {
+            throw DexCountException("Error loading dex file", e)
+        } catch (e: DexDataException) {
+            throw DexCountException("Error loading dex file", e)
+        }
+    }
+
+    override fun close() {
+        raf.close()
+        if (isTemp) {
+            file.delete()
+        }
+    }
+
+    companion object {
+        /**
+         * Extracts a list of {@link DexFile} instances from the given file.
+         *
+         * DexFiles can be extracted either from an Android APK file, or from a raw
+         * {@code classes.dex} file.
+         *
+         * @param file the APK or dex file.
+         * @param dxTimeoutSecs timeout when running Dx in seconds
+         * @return a list of DexFile objects representing data in the given file.
+         */
+        @JvmStatic
+        fun extractDexData(file: File?, dxTimeoutSecs: Int): List<DexFile> {
+            if (file == null || !file.exists()) {
+                return emptyList()
+            }
+
+            // AAR files need special treatment
+            if (file.name.endsWith(".aar")) {
+                return extractDexFromAar(file, dxTimeoutSecs)
+            }
+
+            try {
+                return extractDexFromZip(file)
+            } catch (ignored: ZipException) {
+                // not a zip, no problem
+            }
+
+            return listOf(DexFile(file, false))
+        }
+
+        @JvmStatic
+        private fun extractDexFromAar(file: File, dxTimeoutSecs: Int): List<DexFile> {
+            // unzip classes.jar from the AAR
+            val tempClasses = file.unzip { entries ->
+                val jarFile = entries.find { it.name.matches(Regex("classes\\.jar")) }
+                makeTempFile("classes.jar").also {
+                    jarFile!!.writeTo(it)
```

Comment:
Its fine with me but I just try to avoid `!!` all together. One of the major reasons I am trying to use Kotlin more is to avoid `null`. But yeah, `AAR`s must have `classes.jar`.